source/openqasm/parser/antlr/qasm_parser.py

```diff
@@ -0,0 +1,908 @@
+        # Need to refactor the syntax.
+        if ctx.singleDesignatorType():
+            return SingleDesignatorType(
+                SingleDesignatorTypeName[ctx.singleDesignatorType().getText()],
+                self.visit(ctx.designator()),
+            )
+
+    @span
+    def visitSubroutineDefinition(self, ctx: qasm3Parser.SubroutineDefinitionContext):
+        return SubroutineDefinition(
+            name=ctx.Identifier().getText(),
+            arguments=[
+                self.visit(argument) for argument in ctx.anyTypeArgumentList().anyTypeArgument()
+            ]
+            if ctx.anyTypeArgumentList()
+            else [],
+            return_type=self.visit(ctx.returnSignature().classicalType())
+            if ctx.returnSignature()
+            else None,
+            body=self.visit(ctx.subroutineBlock()),
+        )
+
+    @span
+    def visitAnyTypeArgument(self, ctx: qasm3Parser.AnyTypeArgumentContext):
+        return self.visit(ctx.getChild(0))
+
+    def visitSubroutineBlock(self, ctx: qasm3Parser.SubroutineBlockContext):
+        statements = [self.visit(statement) for statement in ctx.statement()]
+        if ctx.returnStatement():
+            statements.append(self.visit(ctx.returnStatement()))
+
+        return statements
+
+    @span
+    def visitReturnStatement(self, ctx: qasm3Parser.ReturnStatementContext):
+        return ReturnStatement(
+            expression=self.visit(ctx.getChild(1)) if ctx.getChildCount() > 1 else None
+        )
+
+    @span
+    def visitQuantumArgument(self, ctx: qasm3Parser.QuantumArgumentContext):
+        return QuantumArgument(
+            add_span(
+                Qubit(
+                    ctx.Identifier().getText(),
+                ),
+                get_span(ctx.Identifier()),
+            ),
+            self.visit(ctx.designator()) if ctx.designator() else None,
+        )
+
+    @span
+    def visitBranchingStatement(self, ctx: qasm3Parser.BranchingStatementContext):
+        return BranchingStatement(
+            condition=self.visit(ctx.expression()),
+            if_block=self.visit(ctx.programBlock()[0]),
+            else_block=self.visit(ctx.programBlock()[1]) if ctx.getChildCount() > 5 else None,
+        )
+
+    def visitProgramBlock(self, ctx: qasm3Parser.ProgramBlockContext):
+        if ctx.LBRACE():
+            return [self.visit(statement) for statement in list(ctx.getChildren())[1:-1]]
+        else:
+            return [self.visit(ctx.getChild(0))]
+
+    @span
+    def visitControlDirective(self, ctx: qasm3Parser.ControlDirectiveContext):
+        if ctx.returnStatement():
+            return self.visit(ctx.returnStatement())
+        elif ctx.getChild(0).getText() == "break":
+            return BreakStatement()
+        elif ctx.getChild(0).getText() == "continue":
+            return ContinueStatement()
+        elif ctx.getChild(0).getText() == "end":
+            return EndStatement()
```

Comment:
`def visitEndStatement(...)` is necessary because the grammar says
```
statement
    : expressionStatement
    ...
    | endStatement
    ;
```

Reply:
Indeed.